### DISCUSSION TITLE
FAILED TO UPLOAD FILE.

### DISCUSSION THREAD

**Discussion by vivkvv** [plynx-team/plynx#179] [General]
I don't see "Discussion" tab so I will ask here. Is it possible to create a DAG consisting of a few python files? I try to upload a python file using a button on the GUI, but it does not work. I see only the read message "FAILED TO UPLOAD FILE. DID YOU CHOOSE THE RIGHT FILE TYPE?" on the left bottom corner of the screen. I did not find a good explanation of how it works in the documentation. Is there either a video explanation or a series of examples?

Here is a prt of log:

api         | INFO:werkzeug:172.19.0.1 - - [31/Jul/2023 16:44:36] "OPTIONS /plynx/api/v0/upload_file HTTP/1.1" 200 -
api         | INFO:werkzeug:172.19.0.1 - - [31/Jul/2023 16:44:36] "GET /plynx/api/v0/templates/64c7e4bade739cfef47c442e HTTP/1.1" 200 -
api         | INFO:plynx.web.common:<Request 'http://localhost:5005/plynx/api/v0/upload_file' [POST]>
api         | ERROR:plynx.web.common:Traceback (most recent call last):
api         |   File "/usr/local/lib/python3.8/site-packages/plynx-1.11.1-py3.8.egg/plynx/web/common.py", line 155, in decorated
api         |     return f(*args, **kwargs)
api         |   File "/usr/local/lib/python3.8/site-packages/plynx-1.11.1-py3.8.egg/plynx/web/common.py", line 121, in decorated
api         |     return f(*args, **kwargs)
api         |   File "/usr/local/lib/python3.8/site-packages/plynx-1.11.1-py3.8.egg/plynx/web/resource.py", line 82, in upload_file
api         |     output.thumbnail = get_thumbnail(output)
api         |   File "/usr/local/lib/python3.8/site-packages/plynx-1.11.1-py3.8.egg/plynx/utils/thumbnails.py", line 15, in get_thumbnail
api         |     return cls.thumbnail(output)
api         |   File "/usr/local/lib/python3.8/site-packages/plynx-1.11.1-py3.8.egg/plynx/plugins/resources/python/common.py", line 42, in thumbnail
api         |     res = json.load(f)
api         |   File "/usr/local/lib/python3.8/json/__init__.py", line 293, in load
api         |     return loads(fp.read(),
api         |   File "/usr/local/lib/python3.8/json/__init__.py", line 357, in loads
api         |     return _default_decoder.decode(s)
api         |   File "/usr/local/lib/python3.8/json/decoder.py", line 337, in decode
api         |     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
api         |   File "/usr/local/lib/python3.8/json/decoder.py", line 355, in raw_decode
api         |     raise JSONDecodeError("Expecting value", s, err.value) from None
api         | json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
api         |
api         | INFO:werkzeug:172.19.0.1 - - [31/Jul/2023 16:44:36] "POST /plynx/api/v0/upload_file HTTP/1.1" 500 -
api         | INFO:werkzeug:172.19.0.1 - - [31/Jul/2023 16:44:37] "GET /plynx/api/v0/templates/64c7e43ede739cfef47c442d HTTP/1.1" 200 -
api         | INFO:werkzeug:172.19.0.6 - - [31/Jul/2023 16:44:37] "POST /plynx/api/v0/push_worker_state HTTP/1.1" 200 -
api         | INFO:werkzeug:172.19.0.6 - - [31/Jul/2023 16:44:37] "POST /plynx/api/v0/pick_run HTTP/1.1" 200 -
api         | INFO:werkzeug:172.19.0.5 - - [31/Jul/2023 16:44:37] "POST /plynx/api/v0/pick_run HTTP/1.1" 200 -
api         | INFO:werkzeug:172.19.0.5 - - [31/Jul/2023 16:44:37] "POST /plynx/api/v0/push_worker_state HTTP/1.1" 200 -
api         | INFO:werkzeug:172.19.0.1 - - [31/Jul/2023 16:44:37] "GET /plynx/api/v0/templates/64c7e4bade739cfef47c442e HTTP/1.1" 200 -
api         | INFO:werkzeug:172.19.0.1 - - [31/Jul/2023 16:44:38] "GET /plynx/api/v0/templates/64c7e43ede739cfef47c442d HTTP/1.1" 200 -
api         | INFO:werkzeug:172.19.0.6 - - [31/Jul/2023 16:44:38] "POST /plynx/api/v0/push_worker_state HTTP/1.1" 200 -
api         | INFO:werkzeug:172.19.0.6 - - [31/Jul/2023 16:44:38] "POST /plynx/api/v0/pick_run HTTP/1.1" 200 -

**Comment by vivkvv** [comment]
Maybe, the problem is I try to load a file with windows path. I'm not sure

**Comment by khaxis** [comment]
Ok, seems like I know what's happening.
<img width="206" alt="image" src="https://github.com/plynx-team/plynx/assets/2409882/0ab9b5c6-1095-4ff5-b4be-bc1ea8dd65b3">
Are you trying to "Upload Python file"? I think if you click just "Upload File" it should work fine.

The reason this is happening is because of my poor naming. There are two backends, the standard distributed DAG, and Python. They are using different plugins for the "Files". The `Python` plugin used in the default configuration is trying to parse it as a json and that's why there is an error.

How to fix it? Depends on what you are trying to do. Originally in the DAG/bash mode the idea was to do something like this:
<img width="435" alt="image" src="https://github.com/plynx-team/plynx/assets/2409882/6230b8d6-d0ca-4bce-b0b7-b75a73f4544a">
In the Python mode, Instead of a script, I'd use include the code in the Operations themselves (like in the demo on plynx.com)

Does it make sense? May I know a little bit more about your use case and/or expected usage?

**Comment by vivkvv** [comment]
Yes, thanks. But I don't know how to use this file in further graph building.

In my case, I have few Python files and each of them has input data files and makes some work and produces data file(s). I was going to create a DAG and check how it works. I just wanted to check if I can use this tool for my experiments and compare it with Airflow, but I see too many obstacles. It would be good to have a few videos with examples and more details explanations of what to do.

**Reply by khaxis** [reply]
Yes, I see there is a need to write more documentation and record a video with explanation and howto. I will record it later. Meanwhile happy to answer any questions!

Yes airflow is a defacto standard, it has a different use case though. PLynx is more about dynamic DAGs whereas Airflow works with repetitive and stable pipelines.

**Reply by khaxis** [reply]
As of the example above, I think the easiest is to create a bash Operation (**BashJinja2 Operation**) from the standard config.

*Step 1*
<img width="1595" alt="image" src="https://github.com/plynx-team/plynx/assets/2409882/2a139f1e-42bf-464a-be4a-ca0a1ef03219">
```bash
set -e

python {{ inputs.my_script }} \
    --input {{ inputs.input_file }} \
    --output-file {{ outputs.out_file }}
```

*Step 2*
I uploaded a script `example.py` with the following content:
```python
# example.py
import argparse

def copy_file(input_file, output_file):
    try:
        with open(input_file, 'r') as infile:
            content = infile.read()
            with open(output_file, 'w') as outfile:
                outfile.write(content)
        print(f"Content copied from '{input_file}' to '{output_file}' successfully.")
    except FileNotFoundError:
        print(f"Error: Input file '{input_file}' not found.")
    except IOError as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Copy the content from an input file to an output file.")
    parser.add_argument("-i", "--input-file", required=True, help="Path to the input file.")
    parser.add_argument("-o", "--output-file", required=True, help="Path to the output file.")

    args = parser.parse_args()

    input_file = args.input_file
    output_file = args.output_file

    copy_file(input_file, output_file)
```

* Step 3*
Upload a file. Any content would work

*Finally*
<img width="1590" alt="image" src="https://github.com/plynx-team/plynx/assets/2409882/fec623de-9716-425c-9686-135e17485eca">
